From 589397a27759bd650b3674029cb0ef73347c913b Mon Sep 17 00:00:00 2001
From: Dan Streetman <ddstreet@canonical.com>
Date: Wed, 15 Apr 2020 14:40:21 -0400
Subject: [PATCH 1/4] network: change UseGateway= default to UseRoutes= setting
Origin: upstream, https://github.com/systemd/systemd/pull/15443
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1867375

Anyone previously using the UseRoutes=false parameter expected their
dhcp4-provided gateway route to be ignored, as well.  However, with
the introduction of the UseGateway= parameter, this is no longer true.

In order to keep backwards compatibility, this sets the UseGateway=
default value to whatever UseRoutes= has been set to.
---
 man/systemd.network.xml                  | 5 +++--
 src/network/networkd-network-gperf.gperf | 2 +-
 src/network/networkd-network.c           | 5 ++++-
 src/network/networkd-network.h           | 2 +-
 4 files changed, 9 insertions(+), 5 deletions(-)

--- a/man/systemd.network.xml
+++ b/man/systemd.network.xml
@@ -1226,8 +1226,9 @@
         <varlistentry>
           <term><varname>UseGateway=</varname></term>
           <listitem>
-            <para>When true (the default), the gateway will be requested from the DHCP server and added to the
-            routing table with a metric of 1024, and a scope of "link".</para>
+            <para>When true, the gateway will be requested from the DHCP server and added to the routing table with a
+            metric of 1024, and a scope of "link".  When unset, the value specified with <option>UseRoutes=</option>
+            is used.</para>
           </listitem>
         </varlistentry>
         <varlistentry>
--- a/src/network/networkd-network-gperf.gperf
+++ b/src/network/networkd-network-gperf.gperf
@@ -119,7 +119,7 @@ DHCP.UseMTU,
 DHCP.UseHostname,                       config_parse_bool,                              0,                             offsetof(Network, dhcp_use_hostname)
 DHCP.UseDomains,                        config_parse_dhcp_use_domains,                  0,                             offsetof(Network, dhcp_use_domains)
 DHCP.UseRoutes,                         config_parse_bool,                              0,                             offsetof(Network, dhcp_use_routes)
-DHCP.UseGateway,                        config_parse_bool,                              0,                             offsetof(Network, dhcp_use_gateway)
+DHCP.UseGateway,                        config_parse_tristate,                          0,                             offsetof(Network, dhcp_use_gateway)
 DHCP.Anonymize,                         config_parse_bool,                              0,                             offsetof(Network, dhcp_anonymize)
 DHCP.SendHostname,                      config_parse_bool,                              0,                             offsetof(Network, dhcp_send_hostname)
 DHCP.Hostname,                          config_parse_hostname,                          0,                             offsetof(Network, dhcp_hostname)
--- a/src/network/networkd-network.c
+++ b/src/network/networkd-network.c
@@ -211,7 +211,7 @@ static int network_load_one(Manager *man
         network->dhcp_use_dns = true;
         network->dhcp_use_hostname = true;
         network->dhcp_use_routes = true;
-        network->dhcp_use_gateway = true;
+        network->dhcp_use_gateway = -1;
         /* NOTE: this var might be overwriten by network_apply_anonymize_if_set */
         network->dhcp_send_hostname = true;
         network->dhcp_use_domains = DHCP_USE_DOMAINS_YES;
@@ -303,6 +303,9 @@ static int network_load_one(Manager *man
                 network->dhcp_use_mtu = false;
         }
 
+        if (network->dhcp_use_gateway < 0)
+                network->dhcp_use_gateway = network->dhcp_use_routes;
+
         LIST_PREPEND(networks, manager->networks, network);
 
         r = hashmap_ensure_allocated(&manager->networks_by_name, &string_hash_ops);
--- a/src/network/networkd-network.h
+++ b/src/network/networkd-network.h
@@ -149,7 +149,7 @@ struct Network {
         bool dhcp_use_ntp;
         bool dhcp_use_mtu;
         bool dhcp_use_routes;
-        bool dhcp_use_gateway;
+        int dhcp_use_gateway;
         bool dhcp_use_timezone;
         bool rapid_commit;
         bool dhcp_use_hostname;
